progenitor-impl/src/method.rs

```diff
@@ -2224,18 +2337,207 @@ impl Generator {
+                    schemars::schema::SchemaObject {
+                        object: Some(ref mut object),
+                        ..
+                    },
+                ) = schema
+                {
+                    for param in body_params.iter() {
+                        object.properties.remove_entry(&param.api_name);
+                    }
+                };
+                let name = sanitize(
+                    &format!(
+                        "{}-body",
+                        operation.operation_id.as_ref().unwrap(),
+                    ),
+                    Case::Pascal,
+                );
+                let typ =
+                    self.type_space.add_type_with_name(&schema, Some(name))?;
+                OperationParameterType::Type(typ)
+            }
         };
 
-        Ok(Some(OperationParameter {
+        let body_param = OperationParameter {
             name: "body".to_string(),
             api_name: "body".to_string(),
             description: body.description.clone(),
             typ,
             kind: OperationParameterKind::Body(content_type),
-        }))
+        };
+        body_params.push(body_param);
+
+        Ok(body_params)
+    }
+
+    fn get_form_builder(
+        &self,
+        params: &[OperationParameter],
+        form_ident: &proc_macro2::Ident,
+    ) -> Result<Vec<TokenStream>> {
+        let mut form_parts = vec![];
+        let mut form_parts_optional = vec![];
+        let mut build_form_type_id = None;
+        let mut form_data_names = std::collections::HashSet::with_capacity(1);
+
+        params
+            .iter()
+            .filter_map(|param| match (&param.kind, &param.typ) {
+                (
+                    OperationParameterKind::Body(BodyContentType::FormData(
+                        required,
+                    )),
+                    OperationParameterType::FormPart,
+                ) => Some((*required, &param.api_name)),
+                (
+                    OperationParameterKind::Body(BodyContentType::FormData(
+                        true,
+                    )),
+                    OperationParameterType::Type(type_id),
+                ) => {
+                    build_form_type_id = Some(type_id);
+                    None
+                }
+                _ => None,
+            })
+            .for_each(|(required, api_name)| {
+                form_data_names.insert(api_name.clone());
+                let param_ident = format_ident!("{}", api_name);
+                if required {
+                    form_parts.push(quote! { .part(#api_name, #param_ident) });
+                } else {
+                    form_parts_optional.push(quote! {
+                        if let Some(#param_ident) = #param_ident {
+                            #form_ident = #form_ident.part(#api_name, #param_ident);
+                        };
+                    });
+                };
+            });
```

Comment:
Let's get rid of the filter_map with side-effects. Let's just do `for param in params { ` and go from there.